test/e2e_node/pod_status_test.go

```diff
@@ -0,0 +1,140 @@
+		init := "init"
+		regular := "regular"
+
+		podLabels := map[string]string{
+			"test":      "pods-status-phase-test-serial",
+			"namespace": f.Namespace.Name,
+		}
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "initialized-pod",
+				Labels: podLabels,
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyAlways,
+				InitContainers: []v1.Container{
+					{
+						Name:  init,
+						Image: busyboxImage,
+						Command: ExecCommand(init, execCommand{
+							Delay:    30,
+							ExitCode: 0,
+						}),
+					},
+				},
+				Containers: []v1.Container{
+					{
+						Name:  regular,
+						Image: busyboxImage,
+						Command: ExecCommand(regular, execCommand{
+							Delay:    300,
+							ExitCode: 0,
+						}),
+					},
+				},
+			},
+		}
+		preparePod(pod)
+
+		client := e2epod.NewPodClient(f)
+		pod = client.Create(ctx, pod)
+
+		ginkgo.By("Waiting for the pod's status to become Running")
+		err := e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Getting the current pod sandbox ID")
+		rs, _, err := getCRIClient()
+		framework.ExpectNoError(err)
+		sandboxes, err := rs.ListPodSandbox(ctx, &runtimeapi.PodSandboxFilter{
+			LabelSelector: podLabels,
+		})
+		framework.ExpectNoError(err)
+		gomega.Expect(sandboxes).To(gomega.HaveLen(1))
+		podSandboxID := sandboxes[0].Id
+
+		// We need to wait for the pod to be Running before simulating the node reboot,
+		// to avoid any unintended effects from the previous init container state.
+		// Simulate node reboot by restarting the kubelet and the pod sandbox.
+		ginkgo.By("Stopping the kubelet")
+		startKubelet := stopKubelet()
+		gomega.Eventually(ctx, func() bool {
+			return kubeletHealthCheck(kubeletHealthCheckURL)
+		}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeFalseBecause("kubelet should be stopped"))
+
+		ginkgo.By("Stopping the pod sandbox")
+		err = rs.StopPodSandbox(ctx, podSandboxID)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Starting the kubelet")
+		startKubelet()
+		kubeletRestartedTime := time.Now().UTC()
+		gomega.Eventually(ctx, func() bool {
+			return kubeletHealthCheck(kubeletHealthCheckURL)
+		}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be started"))
+
+		ginkgo.By("Waiting for the regular init container to be started after the node reboot")
+		err = e2epod.WaitForPodInitContainerStarted(ctx, f.ClientSet, pod.Namespace, pod.Name, 0, f.Timeouts.PodStart)
+		framework.ExpectNoError(err)
+
+		pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		gomega.Expect(pod.Status.Phase == v1.PodPending).To(gomega.BeTrueBecause("pod should be pending during the execution of the init container after the node reboot"))
+
+		ginkgo.By("Get the logs of the init container after the kubelet restart")
```

Comment:
There is a helper function available that can parse container logs to check if a restart has occurred, so it might be better to rewrite it as follows.

```go
		ginkgo.By("Parse the logs of the pod after the kubelet restart")
		results := parseOutput(ctx, f, pod)
		ginkgo.By("Analyzing results")
		initStartedTimeUnix, err := results.TimeOfLastStart(init)
		framework.ExpectNoError(err)
		initStartedTime := time.UnixMilli(initStartedTimeUnix)
		gomega.Expect(initStartedTime).Should(gomega.BeTemporally(">", kubeletRestartedTime),
			"The logs should contain the most recent occurrence of \"Starting\" output from init container after the kubelet restart")
```

Reply:
Thank you, this looks great.